### PR TITLE
[Placement Group] Support infeasible placement groups for Placement Group.

### PR DESCRIPTION
**Pull request by clay4megtr**
## Why are these changes needed?

For issue: #15416

This pr is just an optimization for strict strategy (like StrictPack StrictSpread) and we will implement the optimization for soft strategy next pr.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Btw, isn't the title misleading? This is basically to support infeasible placement group, am I correct?

**Comment by clay4444** [comment]
> Btw, isn't the title misleading? This is basically to support infeasible placement group, am I correct?

hmmm..  Yeah, but.....  It's also a  scheduling optimization because we will always retry even it's an infeasible pg, right?

**Comment by rkooo567** [comment]
I think "optimization" is too general term. Let me just fix the title. I was just trying to understand the scope of the PR. I will review it by tomorrow! (please ping me again if I dont!)

**Comment by clay4444** [comment]
> I think "optimization" is too general term. Let me just fix the title. I was just trying to understand the scope of the PR. I will review it by tomorrow! (please ping me again if I dont!)

Got it, Thx!

**Comment by clay4444** [comment]
ping~ @rkooo567

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc [hunk 1]
```diff
@@ -284,8 +290,8 @@ void GcsPlacementGroupManager::SchedulePendingPlacementGroups() {
     MarkSchedulingStarted(placement_group_id);
     gcs_placement_group_scheduler_->ScheduleUnplacedBundles(
         placement_group,
-        [this](std::shared_ptr<GcsPlacementGroup> placement_group) {
-          OnPlacementGroupCreationFailed(std::move(placement_group));
+        [this](std::shared_ptr<GcsPlacementGroup> placement_group, bool is_retryable) {
```
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:293]: sure~

File: src/ray/gcs/gcs_server/gcs_resource_scheduler.h [hunk 2]
```diff
@@ -35,7 +35,18 @@ enum SchedulingType {
   SchedulingType_MAX = 4,
 };
 
+// Status of resource scheduling result.
+enum SchedulingResultStatus {
+  // Scheduling failed but retryable.
+  FAILED_BUT_RETRYABLE = 0,
```
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_resource_scheduler.h:41]: Good Idea!

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc [hunk 3]
```diff
@@ -570,6 +576,23 @@ void GcsPlacementGroupManager::OnNodeDead(const NodeID &node_id) {
...
+
+  // Move all the infeasible placement groups to the pending queue so that we can
+  // reschedule them.
+  if (infeasible_placement_groups_.size() > 0) {
+    auto end_it = pending_placement_groups_.end();
+    pending_placement_groups_.insert(end_it, infeasible_placement_groups_.cbegin(),
+                                     infeasible_placement_groups_.cend());
+    infeasible_placement_groups_.clear();
+
+    SchedulePendingPlacementGroups();
```
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:592]: Yeah, thanks for reminding, I think we need to put the `SchedulePendingPlacementGroups()` out of the if code statement, It's safe this way.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc [hunk 4]
```diff
@@ -221,27 +221,33 @@ PlacementGroupID GcsPlacementGroupManager::GetPlacementGroupIDByName(
 }
 
 void GcsPlacementGroupManager::OnPlacementGroupCreationFailed(
-    std::shared_ptr<GcsPlacementGroup> placement_group) {
+    std::shared_ptr<GcsPlacementGroup> placement_group, bool is_retryable) {
```
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc]: updated.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc]: updated

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 5]
```diff
@@ -407,8 +407,9 @@ void GcsServer::InstallEventListeners() {
   gcs_node_manager_->AddNodeAddedListener([this](std::shared_ptr<rpc::GcsNodeInfo> node) {
     // Because a new node has been added, we need to try to schedule the pending
     // placement groups and the pending actors.
+    auto node_id = NodeID::FromBinary(node->node_id());
     gcs_resource_manager_->OnNodeAdd(*node);
-    gcs_placement_group_manager_->SchedulePendingPlacementGroups();
+    gcs_placement_group_manager_->OnNodeAdd(node_id);
```
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_server.cc]: updated.

File: src/ray/gcs/gcs_server/gcs_resource_scheduler.h [hunk 6]
```diff
@@ -35,7 +35,18 @@ enum SchedulingType {
...
+  // Scheduling failed and non-retryable.
+  FAILED_AND_NON_RETRYABLE = 1,
+  // Scheduling successful.
+  SUCCESSFUL = 2,
```
[the remark this answers is not in the document]
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_resource_scheduler.h]: updated.

**Comment by clay4444** [comment]
@rkooo567  sorry, I choose the wrong target repository when I pushed it (You can find an `improve_schedule_process_for_pg` branch in the ray-project),  Can you help to delete it?

**Comment by rkooo567** [comment]
@clay4444 Can you work on test fixes? Also cc @wuisawesome can you take a quick look at the PR?

**Comment by clay4444** [comment]
> @clay4444 Can you work on test fixes? Also cc @wuisawesome can you take a quick look at the PR?

sure~

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc [hunk 7]
```diff
@@ -221,27 +221,33 @@ PlacementGroupID GcsPlacementGroupManager::GetPlacementGroupIDByName(
...
-            state == rpc::PlacementGroupTableData::PENDING ||
-            state == rpc::PlacementGroupTableData::REMOVED)
-      << "State: " << state;
-  if (state == rpc::PlacementGroupTableData::RESCHEDULING) {
-    // NOTE: If a node is dead, the placement group scheduler should try to recover the
-    // group by rescheduling the bundles of the dead node. This should have higher
-    // priority than trying to place other placement groups.
-    pending_placement_groups_.emplace_front(std::move(placement_group));
+
+  if (!is_infeasible) {
```
Comment (wuisawesome) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:228]: Shouldn't this variable be called `is_feasible`?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
One small question.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc (hunk 7 above)
Comment (clay4444) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:228]: Yeah, That make sense to me.

**Review by wuisawesome — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]
LGTM if tests pass!

**Comment by rkooo567** [comment]
Re-run failed rllib test

**Comment by rkooo567** [comment]
Mac wheel build failure is common.

**Pull request merged**